File: frontend/src/api/assets.ts
import { api } from '@/lib/axios';

export interface Asset {
  id: string;
  name: string;
  asset_type_id?: string;
  asset_set_id?: string;
  serial_number?: string;
  assigned_user_id?: string;
  location?: string;
  status_id?: string;
  tags?: string[];
  source?: string;
  
  // Financials
  purchase_price?: number;
  purchase_date?: string;
  vendor?: string;
  order_number?: string;
  warranty_end?: string;
  
  extra?: Record<string, any>;
  created_at?: string;
  updated_at?: string;
  
  // Expanded relations (optional, depending on backend response)
  asset_type?: AssetType;
  status?: AssetStatus;
  asset_set?: AssetSet;
  assigned_user?: User;
}

export interface User {
  id: string;
  email: string;
  full_name?: string;
  is_active: boolean;
  roles: string[];
}

export interface AssetCreate {
  name: string;
  asset_type_id?: string;
  asset_set_id?: string;
  serial_number?: string;
  location?: string;
  status_id?: string;
  assigned_user_id?: string;
  tags?: string[];
  
  // Financials
  purchase_price?: number;
  purchase_date?: string;
  vendor?: string;
  order_number?: string;
  warranty_end?: string;
  
  extra?: Record<string, any>;
}

export interface AssetUpdate {
  name?: string;
  asset_type_id?: string | null;
  asset_set_id?: string | null;
  serial_number?: string | null;
  assigned_user_id?: string | null;
  location?: string | null;
  status?: string;
  tags?: string[];
  
  // Financials
  purchase_price?: number | null;
  purchase_date?: string | null;
  vendor?: string | null;
  order_number?: string | null;
  warranty_end?: string | null;
  
  extra?: Record<string, any>;
}

export interface AssetSet {
  id: string;
  name: string;
  description?: string;
  created_by_id?: string;
  created_at?: string;
}

export interface AssetSetCreate {
  name: string;
  description?: string;
}

export interface AssetSetUpdate {
  name?: string;
  description?: string;
}

export interface AssetType {
  id: string;
  name: string;
  description?: string;
}

export const CustomFieldTarget = {
  asset: "asset",
  user: "user",
} as const;

export type CustomFieldTarget = typeof CustomFieldTarget[keyof typeof CustomFieldTarget];

export const CustomFieldType = {
  string: "string",
  integer: "integer",
  boolean: "boolean",
  date: "date",
  enum: "enum",
  reference: "reference",
} as const;

export type CustomFieldType = typeof CustomFieldType[keyof typeof CustomFieldType];

export interface CustomFieldDefinition {
  id: string;
  target: CustomFieldTarget;
  key: string;
  label: string;
  field_type: CustomFieldType;
  asset_type_id?: string;
  asset_set_id?: string;
  required: boolean;
  order: number;
  config?: Record<string, any>;
}

export interface CustomFieldDefinitionCreate {
  label: string;
  key: string;
  target: CustomFieldTarget;
  field_type: CustomFieldType;
  asset_type_id?: string;
  asset_set_id?: string;
  required?: boolean;
  order?: number;
  config?: Record<string, any>;
}

export interface CustomFieldDefinitionUpdate {
  label?: string;
  field_type?: CustomFieldType;
  asset_type_id?: string | null;
  asset_set_id?: string | null;
  required?: boolean;
  order?: number;
  config?: Record<string, any>;
}

export const getCustomFields = async () => {
  const { data } = await api.get<CustomFieldDefinition[]>('/custom-fields/');
  return data;
};

export const createCustomField = async (payload: CustomFieldDefinitionCreate) => {
  const { data } = await api.post<CustomFieldDefinition>('/custom-fields/', payload);
  return data;
};

export const updateCustomField = async (id: string, payload: CustomFieldDefinitionUpdate) => {
  const { data } = await api.patch<CustomFieldDefinition>(`/custom-fields/${id}`, payload);
  return data;
};

export const deleteCustomField = async (id: string) => {
  await api.delete(`/custom-fields/${id}`);
};


export interface AssetStatus {
  id: string;
  name: string;
  description?: string;
  is_default: boolean;
  color?: string; // If backend supports it, otherwise UI mapping
}

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface GetAssetsParams {
  page?: number;
  size?: number;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
  search?: string;
  asset_set_id?: string;
  assigned_user_id?: string;
  unassigned_set?: boolean;
  status_id?: string;
  asset_type_id?: string;
}

export interface AuditLog {
  id: string;
  entity_type: string;
  entity_id: string;
  action: string;
  changes: Record<string, any>;
  user_id?: string;
  timestamp: string;
  user_name?: string;
  entity_name?: string;
  origin?: string;
}

export const getAssetHistory = async (assetId: string) => {
  const { data } = await api.get<AuditLog[]>('/audit-logs/', {
    params: {
      entity_type: 'asset',
      entity_id: assetId
    }
  });
  return data;
};

export const getAsset = async (id: string) => {
  const { data } = await api.get<Asset>(`/assets/${id}`);
  return data;
};

export const getAssets = async (params: GetAssetsParams = {}) => {
  const { data } = await api.get<PaginatedResponse<Asset>>('/assets/', { params });
  return data;
};

export const createAsset = async (payload: AssetCreate) => {
  const { data } = await api.post<Asset>('/assets/', payload);
  return data;
};

export const updateAsset = async (id: string, payload: AssetUpdate) => {
  const { data } = await api.patch<Asset>(`/assets/${id}`, payload);
  return data;
};

export const deleteAsset = async (id: string) => {
  await api.delete(`/assets/${id}`);
};

export const bulkDeleteAssets = async (assetIds: string[]) => {
  const { data } = await api.post<{ deleted_count: number }>('/assets/bulk-delete', { asset_ids: assetIds });
  return data;
};

export const getAssetSets = async () => {
  const { data } = await api.get<AssetSet[]>('/asset-sets/');
  return data;
};

export const createAssetSet = async (payload: AssetSetCreate) => {
  const { data } = await api.post<AssetSet>('/asset-sets/', payload);
  return data;
};

export const updateAssetSet = async (id: string, payload: AssetSetUpdate) => {
  const { data } = await api.patch<AssetSet>(`/asset-sets/${id}`, payload);
  return data;
};

export const deleteAssetSet = async (id: string) => {
  await api.delete(`/asset-sets/${id}`);
};

export const getAssetTypes = async () => {
  const { data } = await api.get<AssetType[]>('/asset-types/');
  return data;
};

export const getAssetStatuses = async () => {
  const { data } = await api.get<AssetStatus[]>('/asset-statuses/');
  return data;
};

export const createAssetStatus = async (payload: { key: string; label: string }) => {
  const { data } = await api.post<AssetStatus>('/asset-statuses/', payload);
  return data;
};

export const updateAssetStatus = async (key: string, payload: { label?: string }) => {
  const { data } = await api.patch<AssetStatus>(`/asset-statuses/${key}`, payload);
  return data;
};

export const deleteAssetStatus = async (key: string) => {
  await api.delete(`/asset-statuses/${key}`);
};

export const createAssetType = async (payload: { name: string; description?: string }) => {
  const { data } = await api.post<AssetType>('/asset-types/', payload);
  return data;
};

export const updateAssetType = async (id: string, payload: { name?: string; description?: string }) => {
  const { data } = await api.patch<AssetType>(`/asset-types/${id}`, payload);
  return data;
};

export const deleteAssetType = async (id: string) => {
  await api.delete(`/asset-types/${id}`);
};

File: it-asset-manager/app/api/routers/assets.py
from typing import List, Optional
from uuid import UUID
import logging

from fastapi import APIRouter, Depends, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.db import get_db
from app.core.deps import get_current_user, require_permission
from app.schemas.asset import AssetCreate, AssetUpdate, AssetRead, BulkDeleteAssetsRequest, BulkDeleteAssetsResponse
from app.schemas.common import PaginatedResponse
from app.services import asset_service
from app.core.logging import log_activity
from app.core.exceptions import PermissionDenied

router = APIRouter()
logger = logging.getLogger(__name__)

@router.get("/", response_model=PaginatedResponse[AssetRead])
@log_activity
async def list_assets(
    page: int = 1,
    size: int = 50,
    sort_by: str = "created_at",
    sort_order: str = "desc",
    search: Optional[str] = None,
    asset_set_id: Optional[UUID] = None,
    assigned_user_id: Optional[UUID] = None,
    unassigned_set: bool = False,
    status_id: Optional[UUID] = None,
    asset_type_id: Optional[UUID] = None,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(require_permission("asset:read")),
):
    # Enforce data scoping
    if not current_user.has_permission("asset:view_all"):
        assigned_user_id = current_user.id
        
    # Debug log
    if assigned_user_id:
        logger.debug(f"Filtering assets by assigned_user_id: {assigned_user_id}")
    else:
        logger.debug("No assigned_user_id filter")

    return await asset_service.list_assets(
        db=db,
        page=page,
        size=size,
        sort_by=sort_by,
        sort_order=sort_order,
        search=search,
        asset_set_id=asset_set_id,
        assigned_user_id=assigned_user_id,
        unassigned_set=unassigned_set,
        status_id=status_id,
        asset_type_id=asset_type_id
    )

@router.post("/", response_model=AssetRead, status_code=status.HTTP_201_CREATED)
@log_activity
async def create_asset(
    payload: AssetCreate,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(require_permission("asset:create")),
):
    return await asset_service.create_asset(db, payload, user_id=current_user.id)

@router.post("/bulk-delete", response_model=BulkDeleteAssetsResponse)
@log_activity
async def bulk_delete_assets(
    payload: BulkDeleteAssetsRequest,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(require_permission("asset:delete")),
):
    count = await asset_service.bulk_delete_assets(db, asset_ids=payload.asset_ids, user_id=current_user.id)
    return BulkDeleteAssetsResponse(deleted_count=count)

@router.get("/{asset_id}", response_model=AssetRead)
@log_activity
async def get_asset(
    asset_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(require_permission("asset:read")),
):
    asset = await asset_service.get_asset(db, asset_id)
    
    # Enforce scoping
    if not current_user.has_permission("asset:view_all"):
        if asset.assigned_user_id != current_user.id:
            raise PermissionDenied("Not authorized to view this asset")
            
    return asset

@router.patch("/{asset_id}", response_model=AssetRead)
@log_activity
async def update_asset(
    asset_id: UUID,
    payload: AssetUpdate,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(require_permission("asset:update")),
):
    return await asset_service.update_asset(db, asset_id, payload, user_id=current_user.id)

@router.delete("/{asset_id}", status_code=204)
@log_activity
async def delete_asset(
    asset_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(require_permission("asset:delete")),
):
    await asset_service.delete_asset(db, asset_id, user_id=current_user.id)
    return None
